### ISSUE TITLE
Add --bb-edn or --config option to specify other bb.edn file 

### ISSUE THREAD

**Issue by borkdude** [babashka/babashka#1110]
The other bb.edn (or bb.ci.edn, etc) file can live in another directory.
Relative paths in `:paths` should be resolved against the explicitly provided `--config`'s location.

In the future we can also add `--context` for setting the directory to which we should resolve relative dirs againt, but we can default to the explicit `--config` option when given, or the current working directory otherwise (like it is now).

See discussion in https://github.com/babashka/babashka/discussions/869

/cc @wilkerlucio @cldwalker @thiagokokada @lispyclouds

**Comment by cldwalker** [comment]
I'd be in favor of suggested behavior above. The two use cases I have are:
* Given an explicit config file, run a babashka executable with relative paths resolved to explicit file e.g. [bb-github-repo](https://github.com/cldwalker/bb-clis/blob/master/bin/bb-github-repo) as an executable
* Given an explicit config file, run a babashka task with relative paths resolved to explicit file e.g. `bb --config CONFIG run FOO`

**Comment by borkdude** [comment]
Also see #1103 which is somewhat related to invoking tasks from another tasks file, but with merge behavior.

**Comment by cldwalker** [comment]
That does look helpful and related but it's different enough that I don't think it address my uses cases.  Since #1103 is task specific it doesn't address the executable use case. For the task use case, #1103 requires a local bb.edn in order to merge in another bb.edn. I'm looking to invoke another bb.edn from the commandline

**Comment by borkdude** [comment]
@cldwalker Agreed. I hope to look into this issue soon (but feel free to experiment).

**Comment by borkdude** [comment]
Should work now in branch `other-bb-edn`. Still have to write some tests and docs, but with the manual tests I did it worked.

The argument is `--config other-bb.edn`.  

macOS: https://24985-201467090-gh.circle-artifacts.com/0/release/babashka-0.7.1-SNAPSHOT-macos-amd64.tar.gz
linux: https://24986-201467090-gh.circle-artifacts.com/0/release/babashka-0.7.1-SNAPSHOT-linux-amd64-static.tar.gz
Windows:

**Comment by borkdude** [comment]
Merged to master.

**Comment by borkdude** [comment]
Btw I also implemented `--deps-root` for explicitly setting the dependency root for relative paths.

**Comment by borkdude** [comment]
You can now install a dev release: https://github.com/babashka/babashka-dev-builds
This should hopefully help getting this tested!

**Comment by cldwalker** [comment]
@borkdude Tested --config and it works great for my tasks use case! I ended up not using it for the executable use case as `#!/usr/bin/env bb --config PATH` wouldn't be an executable I could share with others. I used an explicit `deps/add-deps` which works well enough there. I haven't tried `--deps-root` as I didn't have a use for it yet. Thanks!

**Comment by borkdude** [comment]
I guess `#!/usr/bin/env bb --config PATH` could be a bash wrapper script you could share, if your project consisted of many namespaces and wanted to use declarative deps, but yeah, `add-deps` is nice for keeping the shared artifact a single file. I wonder if it would be useful to have something like the `clojure -Ttool` thing. E.g we could have something like:


```
babashka install :git/url https://github.com/borkdude/http-server :git/sha <> :as http-server
```

which will make a `~/.babashka/bin/http-server` bash wrapper which does the config thing.

**Comment by thiagokokada** [comment]
Just a reminder that `env` doesn't support parameters unless the `-S` flag is passed. This is not a POSIX extension, and [it was not supported by coreutils until version 8.3.0](https://unix.stackexchange.com/a/477651) (that is kinda recent, Ubuntu 19.04 was the first one to include it).

AFAIK, `-S` works in BSDs (not that it matters for babashka) and macOS too, so if supporting old Linux distros is not an issue it is a safe flag to use.

TL;DR: shebang should be `#!/usr/bin/env -S bb --config PATH` unless you have an old Linux distro.

**Comment by borkdude** [comment]
Yes, I intending to replace what @cldwalker wrote with a bash wrapper script:

```
#!/usr/bin/env bash

bb --config ...
```

**Comment by thiagokokada** [comment]
> Should work now in branch `other-bb-edn`. Still have to write some tests and docs, but with the manual tests I did it worked.
> 
> The argument is `--config other-bb.edn`. Paths are resolved relative to the other bb.edn.
> 
> macOS: https://24985-201467090-gh.circle-artifacts.com/0/release/babashka-0.7.1-SNAPSHOT-macos-amd64.tar.gz linux: https://24986-201467090-gh.circle-artifacts.com/0/release/babashka-0.7.1-SNAPSHOT-linux-amd64-static.tar.gz
> 
> Please test!

Seems to work fine for my use case too (tested only with `--config`, not with `--deps-root`.

**Comment by borkdude** [comment]
So `--config foo/bar/bb.edn` picks up `foo/bar` as the deps root. If you would have `{:paths ["src"]}` in your `bb.edn` then `foo/bar/src` would be added to the classpath. I'm not sure if this was the right default but since it can be configured, there's always a way to make it do what you want.

**Comment by cldwalker** [comment]
> I guess `#!/usr/bin/env bb --config PATH` could be a bash wrapper script you could share, if your project consisted of many namespaces and wanted to use declarative deps, but yeah, `add-deps` is nice for keeping the shared artifact a single file. I wonder if it would be useful to have something like the `clojure -Ttool` thing. E.g we could have something like:
> 
> ```
> bb add-tool :git/url https://github.com/borkdude/http-server :git/sha <> :as http-server
> ```
> 
> which will make a `~/.babashka/tools/http-server` bash wrapper which does the config thing.
> 
> You would then put `~/.babashka/tools` in your `PATH`

Having a consistent directory that babashka installs tools/commands/executables/scripts to `PATH` could be a powerful and useful way to share babashka scripts. In my case, I have about a dozen+ in one repository at https://github.com/cldwalker/bb-clis/tree/master/bin. A different name than tools could avoid confusion if it doesn't behave like clojure CLI tools

**Comment by thiagokokada** [comment]
> Should work now in branch `other-bb-edn`. Still have to write some tests and docs, but with the manual tests I did it worked.
> 
> The argument is `--config other-bb.edn`. Paths are resolved relative to the other bb.edn.
> 
> macOS: https://24985-201467090-gh.circle-artifacts.com/0/release/babashka-0.7.1-SNAPSHOT-macos-amd64.tar.gz linux: https://24986-201467090-gh.circle-artifacts.com/0/release/babashka-0.7.1-SNAPSHOT-linux-amd64-static.tar.gz
> 
> Please test!

Seems this branch broke something (this code used to work in babashka 0.7.0):

```
----- Error --------------------------------------------------------------------
Type:     java.lang.NullPointerException
Location: /home/thiagoko/dev/nu/nucli.bb/bin/nu:53:1

----- Context ------------------------------------------------------------------
49:        (apply (partial fs/path nucli-bb-dir))
50:        str))
51:
52: (load-file (path-from-nucli-bb-dir "nu.clj"))
53: (deps/add-deps (-> "bb.edn" path-from-nucli-bb-dir slurp edn/read-string))
    ^---
54: (cp/add-classpath (path-from-nucli-bb-dir "src"))
55:
56: ;; Start nucli.bb
57:
58: (require '[nucli.main :as nucli])

----- Stack trace --------------------------------------------------------------
babashka.fs/as-path         - <built-in>
babashka.fs/absolutize      - <built-in>
babashka.impl.deps/add-deps - <built-in>
user                        - /home/thiagoko/dev/nu/nucli.bb/bin/nu:53:1
```

Running the same script inside the directory that the `bb.edn` file is located works.

**Comment by borkdude** [comment]
@thiagokokada Could you make a small repro that works with 0.7.0 but not with master?

**Comment by thiagokokada** [comment]
@borkdude Try this one.

[bb-bug.zip](https://github.com/babashka/babashka/files/7781301/bb-bug.zip)

Extract it and execute the `bin/script` file **outside** of the directory where the `bb.edn` file is located.

On `0.7.0` I get:
```
$ ~/bb-bug/bin/script
Hello World
```

On `0.7.1-SNAPSHOT` I get:

```
$ ~/bb-bug/bin/script
----- Error --------------------------------------------------------------------
Type:     java.lang.NullPointerException
Location: /home/thiagoko/bb-bug/bin/script:23:1

----- Context ------------------------------------------------------------------
19:   (->> paths
20:        (apply (partial fs/path script-dir))
21:        str))
22:
23: (deps/add-deps (-> "bb.edn" path-from-script-dir slurp edn/read-string))
    ^---
24: (cp/add-classpath (path-from-script-dir "src"))
25:
26: (println "Hello World")
27:
28: ;; vim: set ft=clojure:

----- Stack trace --------------------------------------------------------------
babashka.fs/as-path         - <built-in>
babashka.fs/absolutize      - <built-in>
babashka.impl.deps/add-deps - <built-in>
user                        - /home/thiagoko/bb-bug/bin/script:23:1
```

Remember to set `BABASHKA_BIN` (inside `bin/script` file) to the PATH where your babashka binary is located.

**Comment by borkdude** [comment]
Fixed on master
